rust/src/builders/tx_builder.rs

```diff
@@ -3086,6 +3086,47 @@ mod tests {
         assert!(input_total >= Value::new(&tx_builder.min_fee().unwrap().checked_add(&to_bignum(COST)).unwrap()));
     }
 
+    #[test]
+    fn tx_builder_cip2_random_improve_on_real_data() {
```

Comment:
Note that this test doesn't always fail due to the randomness. When it fails, it returns as follow:

```
failures:

---- builders::tx_builder::tests::tx_builder_cip2_random_improve_on_real_data stdout ----
thread 'builders::tx_builder::tests::tx_builder_cip2_random_improve_on_real_data' panicked at 'Input TransactionHash([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]) duplicated', src/builders/tx_builder.rs:3122:17


failures:
    builders::tx_builder::tests::tx_builder_cip2_random_improve_on_real_data

test result: FAILED. 165 passed; 1 failed; 1 ignored; 0 measured; 0 filtered out; finished in 4.30s

error: test failed, to rerun pass '--lib'
```

Reply:
It is a fifty-fifty chance. It fails when 10 ADA is selected first.

Reply:
We should avoid tests that fail randomly. I would feel more comfortable if we changed the function to instead take in a seed to use for the randomness so that by default it's random, but given a seed it will provide a deterministic result. It would help users in some cases and also help make our tests deterministic.

Reply:
With the fix it's 100% passed. However, I can remove this and test with a seed in [here](https://github.com/dcSpark/cardano-multiplatform-lib/pull/50/files#diff-f9e91e0f5b08343e5a3f64238729812bc8daf2f2beabc959064832b61d0d7d05R3155).

Reply:
I created PR #51 to upgrade `rand` and `rand_chacha`. I'll make the change after that is merged. @SebastienGllmt